wcコマンドを作成

samovar1214 opened the pull request.

samovar1214 committed:
wcコマンドを作成

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,58 @@
+else
+  name, counts = file_contents.first
+  puts format_output(name, counts, options)
+end
```
```ruby
file_contents.each do |name, counts|
  puts format_output(name, counts, options)
  total_counts[:lines] += counts[:lines]
  total_counts[:words] += counts[:words]
  total_counts[:bytes] += counts[:bytes]
end
puts format_output('total', total_counts, options) file_contents.size > 1
```

のように、合計表示だけ分岐するようにしたほうが、計算が少し無駄にはなりますが、わかりやすいかなと思いました。どうでしょうか。

samovar1214 replied:
おお～！こちらの方が分岐が「合計を表示するかどうか」だけになって分かりやすいです…！修正します✨

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,58 @@
+options = ARGV.getopts('lwc')
+files   = ARGV
+
+file_contents = if files.empty?
+                  [read_stdin]
+                else
+                  files.map { |file| read_file(file) }
+                end
```
標準入力との違いをうまく吸収できていますね 👍

samovar1214 replied:
ありがとうございます！🥹

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,58 @@
+
+def format_output(name, counts, options)
+  no_option = !options['l'] && !options['w'] && !options['c']
+  fields = [[:lines, 'l'], [:words, 'w'], [:bytes, 'c']]
```
ここは `OPTION_ATTRIBUTE_MAPPING = ...` みたいな感じで定数でもっておくとよいかなと思いました。ここの整合性が崩れることはバグにつながるので、明示的に目につくところに置いたほうがよさそうです。

samovar1214 replied:
安全になるし読みやすくなりますね💡`fields = ...` の部分を定数`OPTION_ATTRIBUTE_MAPPING`で持つように修正します！

yoshitsugu commented on 05.wc/wc.rb:
```diff
@@ -0,0 +1,58 @@
+def count_content(content)
+  {
+    lines: content.lines.size,
+    words: content.split(/\s+/).size,
```
```


test
```

のように空行ではじまるファイルの場合にここの計算が組み込みのwcと異なるものになりそうです。

samovar1214 replied:
たしかにこれだと先頭や末尾にある空行も単語の1つとしてカウントしてしまうんですね！💦取り除く方法を調べて修正します。

yoshitsugu reviewed (commented):
コメントしました！

samovar1214 committed:
単語数のカウント処理を修正

空行や先頭・末尾の空白も単語数としてカウントされており、組み込みのwcコマンドと計算が異なっていたため、stripメソッドで不要な空の要素を取り除くように変更しました。

samovar1214 committed:
オプションの対応表を定数化

可読性と保守性を向上させるため、オプションと出力項目の対応が格納されたfieldsをOPTION_ATTRIBUTE_MAPPINGとして定数化しました。

samovar1214 committed:
totalの分岐を変更

合計値の計算は常に実施しその表示のみ分岐させることで、簡潔で読みやすくしました。

yoshitsugu reviewed (approved):
OKです！
